jms/src/main/scala/akka/stream/alpakka/jms/Jms.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.alpakka.jms
+
+import javax.jms
+import javax.jms._
+
+import akka.stream.stage.{ GraphStageLogic, StageLogging }
+
+trait JmsSettings {
+  def connectionFactory: ConnectionFactory
+  def destination: Option[Destination]
+  def credentials: Option[Credentials]
+}
+
+sealed trait Destination
+case class Topic(name: String) extends Destination
+case class Queue(name: String) extends Destination
+
+object JmsSourceSettings {
+
+  def create(connectionFactory: ConnectionFactory) = JmsSourceSettings(connectionFactory)
+
+}
+
+case class JmsSourceSettings(connectionFactory: ConnectionFactory,
+                             destination: Option[Destination] = None,
+                             credentials: Option[Credentials] = None,
+                             bufferSize: Int = 100)
+    extends JmsSettings {
+  def withCredential(credentials: Credentials) = copy(credentials = Some(credentials))
+  def withBufferSize(size: Int) = copy(bufferSize = size)
+  def withQueue(name: String) = copy(destination = Some(Queue(name)))
+  def withTopic(name: String) = copy(destination = Some(Topic(name)))
+}
+
+object JmsSinkSettings {
+
+  def create(connectionFactory: ConnectionFactory) = JmsSinkSettings(connectionFactory)
+
+}
+
+case class JmsSinkSettings(connectionFactory: ConnectionFactory,
+                           destination: Option[Destination] = None,
+                           credentials: Option[Credentials] = None)
+    extends JmsSettings {
+  def withCredential(credentials: Credentials) = copy(credentials = Some(credentials))
+  def withQueue(name: String) = copy(destination = Some(Queue(name)))
+  def withTopic(name: String) = copy(destination = Some(Topic(name)))
+}
+
+case class Credentials(username: String, password: String)
+
+trait JmsConnector { this: GraphStageLogic with StageLogging =>
+
+  private[jms] var jmsConnection: Option[Connection] = None
+  private[jms] var jmsSession: Option[Session] = None
+  private[jms] var jmsDestination: Option[jms.Destination] = None
+
+  def jmsSettings: JmsSettings
+
+  def fail = getAsyncCallback[JMSException](e => failStage(e))
+
+  def openQueue(): Unit = {
+    val factory: ConnectionFactory = jmsSettings.connectionFactory
+    try {
+      jmsConnection = Option(jmsSettings.credentials match {
+        case Some(Credentials(username, password)) => factory.createConnection(username, password)
+        case _ => factory.createConnection()
+      })
```

Comment:
It's a little bit messy as there is no predefined API way to do it, but you can get it from the underlying:

```
val blockingDispatcher = materializer match {
        case m: ActorMaterializer => m.system.dispatchers.lookup("akka.stream.default-blocking-io-dispatcher")
        case x => throw new IllegalArgumentException(s"Stage only works with the ActorMaterializer, was: $x")
      }
```

Just running the entire stage on the blocking dispatcher is easier and is done by overriding `initialAttributes` returning `ActorAttributes.Dispatcher("akka.stream.default-blocking-io-dispatcher")` on the `GraphStage`.